source/precalculus/exercises/outcomes/EQ/EQ7/generator.sage

```diff
@@ -1,10 +1,81 @@
+load("../sage/common.sage")
+
 class Generator(BaseGenerator):
-    def data(self):
+  def data(self):
+
+    #Task 1
+    roots = [-6..6]
+    roots.extend(TBILPrecal.small_rationals(numerators=[-3..3],denominators=[2,3],dictionary=False,length=3))
+    r1,r2=sample(roots,2)
+    d1=r1.denominator()
+    d2=r2.denominator()
+    quadratic_inequality = CheckIt.shuffled_inequality(*[a*x^b for (a,b) in expand(d1*d2*(x-r1)*(x-r2)).coefficients()])
+    partition_points=sorted([r1,r2])
+
+    quadratic_intervals=TBILPrecal.intervals_from_inequality(quadratic_inequality,partition_points)
+
+    #Task 2
+    partition_points=[]
+    equation_type=choice(["linear", "quadratic"])
+
+    if equation_type=="linear":
+      #Ensure a=d
+
+      a=choice([0..6])
+      d=a
+      c=choice([-6..-1,1..6])
+      e=choice([-6..-1,1..6])
+      f=choice([-6..-1,1..6])
+      b=(f-1)*a*(f-c)+e*(c-f+1)
+      partition_points = [1-f]
+
+    else:
+      #Quadratic
+      e=choice([-6..-1,1..6])
+      r1,r2 = sample([-5..-1,1..5],2)
+      #Pre-empt f=c causing a divide by zero error below
+      c= choice([a for a in [-5..-1,1..5] if a!=r1 and a!=r2 and (e==1 or a!= r1*r2/(1-e))])
+      b=1
+      f=r1*r2+e*c
+      a=(-r1-r2+e-c-b)/(f-c)
+      d=a-1
+      partition_points = [r1,r2]
+
+    undefined_points=[-c,-f]
+    #Prevent duplicates
+    partition_points.extend([x for x in undefined_points if x not in partition_points])
+    partition_points.sort()
+
+    LHS = (expand((a*x+b)*a.denominator())).mul(1/(x+c),hold=True)
+    RHS = (expand((d*x+e)*a.denominator())).mul(1/(x+f),hold=True)
+    ineq=choice(["<","<=",">",">="])
+    if ineq == "<":
+      rational_inequality = LHS < RHS
+    elif ineq == "<=":
+      rational_inequality = LHS <= RHS
+    elif ineq == ">":
+      rational_inequality = LHS > RHS
+    elif ineq == ">=":
+      rational_inequality = LHS >= RHS
```

Comment:
I couldn't see a clean way to use `CheckIt.shuffled_inequality()` here, as all I have to play with are two rational functions, and I just want $P(x)<Q(x)$ or similar. The `shuffled_inequality()` kept doing silly things like $P(x)+Q(x)>0 $which I didn't want.

Reply:
After stewing on this over the weekend, thought up a cleaner fix implemented in [ac06620](https://github.com/TeamBasedInquiryLearning/library/pull/216/commits/ac0662031f5bb64b7191cbbb589ee77c321cf1dd) using the `operator` module.